### ISSUE TITLE
Galaxy timelapse frame export fails in windows build

### ISSUE THREAD

**Issue by eliasdoehne** [eliasdoehne/stellaris-dashboard#98]
Using the windows build from this workflow run https://github.com/eliasdoehne/stellaris-dashboard/actions/runs/3848208751 with [unitednationsofearth7_1529854637.zip](https://github.com/eliasdoehne/stellaris-dashboard/files/10362443/unitednationsofearth7_1529854637.zip)
 gives this exception

```
MainProcess - 2023-01-06 20:00:28,583 - stellarisdashboard.dashboard_app.timelapse_exporter - INFO - Exporting timelapse frames to C:\Users\...\stellarisdashboard-windows-latest\output\galaxy-timelapse\unitednationsofearth7_1529854637-2023-01-06-200028
  0%|                                                                                            | 0/7 [00:00<?, ?it/s]
MainProcess - 2023-01-06 20:00:28,986 - stellarisdashboard.dashboard_app - ERROR - Exception on /timeline/_dash-update-component [POST]
Traceback (most recent call last):
  File "flask\app.py", line 2525, in wsgi_app
  File "flask\app.py", line 1822, in full_dispatch_request
  File "flask\app.py", line 1820, in full_dispatch_request
  File "flask\app.py", line 1796, in dispatch_request
  File "dash\dash.py", line 1274, in dispatch
    ctx.run(
  File "dash\_callback.py", line 440, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
  File "stellarisdashboard\dashboard_app\graph_ledger.py", line 309, in trigger_timeline_export
  File "stellarisdashboard\dashboard_app\timelapse_exporter.py", line 71, in create_timelapse
  File "PIL\Image.py", line 2431, in save
  File "PIL\PngImagePlugin.py", line 1396, in _save
TypeError: unsupported operand type(s) for /: 'NoneType' and 'float'
```

![](https://user-images.githubusercontent.com/3602589/211081436-0874a8c6-dcf1-44e2-8a9c-88e772bd7f0d.png)

**Comment by eliasdoehne** [comment]
actually I get the same error with that db when running from the current master directly
```
Traceback (most recent call last):
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\flask\app.py", line 2525, in wsgi_app
    response = self.full_dispatch_request()
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\flask\app.py", line 1822, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\flask\app.py", line 1820, in full_dispatch_request
    rv = self.dispatch_request()
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\flask\app.py", line 1796, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\dash\dash.py", line 1274, in dispatch
    ctx.run(
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\dash\_callback.py", line 440, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\stellarisdashboard\dashboard_app\graph_ledger.py", line 309, in trigger_timeline_export
    te.create_timelapse(
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\stellarisdashboard\dashboard_app\timelapse_exporter.py", line 71, in create_timelapse
    frame.save(
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\PIL\Image.py", line 2431, in save
    save_handler(self, fp, filename)
  File "C:\Users\elias\PycharmProjects\stellaris-dashboard\env\lib\site-packages\PIL\PngImagePlugin.py", line 1396, in _save
    o32(int(dpi[0] / 0.0254 + 0.5)),
TypeError: unsupported operand type(s) for /: 'NoneType' and 'float'
```

so it's probably not related to the build

**Comment by quantumrook** [comment]
Not sure if you're still interested in resolving this, but if you are:

### TLDR

It appears that the `dpi` parameter for `graph_ledger.py:trigger_timeline_export(...)` was not being checked for `None` in the case where the user did not interact with the html input control for DPI.

### Details

Turning on logging in the config shows that the information being passed to the `TimelapseExporter's Constructor` ([quick link](https://github.com/eliasdoehne/stellaris-dashboard/blob/f0316d4a4b0f7346aa09cf955c2c4955c25e120d/stellarisdashboard/dashboard_app/timelapse_exporter.py#L27)) is `None`:

> 2025-09-13 13:35:17,840 - stellarisdashboard.dashboard_app.graph_ledger - INFO - start_date='2200.01.01', end_date='2200.02.01', step_days=120, frame_time_ms=100, export_gif=True, export_webp=False, export_frames=True, square_aspect_ratio=False, dpi=None

The core of the issue seems to come from `graph_ledger.py:270` ([quick link](https://github.com/eliasdoehne/stellaris-dashboard/blob/f0316d4a4b0f7346aa09cf955c2c4955c25e120d/stellarisdashboard/dashboard_app/graph_ledger.py#L270)) where the `dpi` parameter is not checked like the others:

```py
    start_date = start_date or TIMELAPSE_DEFAULT_START
    if not end_date:
        with datamodel.get_db_session(game_id) as session:
            end_date = datamodel.days_to_date(utils.get_most_recent_date(session))
    step_days = step_days or TIMELAPSE_DEFAULT_STEP
    frame_time_ms = frame_time_ms or TIMELAPSE_DEFAULT_FRAME_TIME
```

after adding the line,

```py
    dpi = dpi or TIMELAPSE_DEFAULT_DPI
```

rebuilding, and testing; the export works as originally expected.

Log output:

> 2025-09-13 15:02:06,780 - stellarisdashboard.dashboard_app.graph_ledger - INFO - start_date='2200.01.01', end_date='2200.02.01', step_days=120, frame_time_ms=100, export_gif=True, export_webp=False, export_frames=True, square_aspect_ratio=False, dpi=100

### Note

During the testing process, I accidentally set the dpi to `1` in the browser view (while investigating using developer tools) and managed to cause a different exception to be thrown. The gist is there appears to be some assumed minimum possible value for dpi somewhere in `matplotlib`, which boils up as:

> 2025-09-13 15:12:53,498 - stellarisdashboard.dashboard_app - ERROR - Exception on /timeline/_dash-update-component [POST]

```
Traceback (most recent call last):
  File "flask\app.py", line 2529, in wsgi_app
  File "flask\app.py", line 1825, in full_dispatch_request
  File "flask\app.py", line 1823, in full_dispatch_request
  File "flask\app.py", line 1799, in dispatch_request
  File "dash\dash.py", line 1310, in dispatch
    ctx.run(
  File "dash\_callback.py", line 442, in add_context
    output_value = func(*func_args, **func_kwargs)  # %% callback invoked %%
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "stellarisdashboard\dashboard_app\graph_ledger.py", line 326, in trigger_timeline_export
  File "stellarisdashboard\dashboard_app\timelapse_exporter.py", line 66, in create_timelapse
  File "stellarisdashboard\dashboard_app\timelapse_exporter.py", line 172, in draw_frame
  File "matplotlib\figure.py", line 3388, in savefig
  File "matplotlib\backend_bases.py", line 2218, in print_figure
  File "matplotlib\backend_bases.py", line 2068, in <lambda>
  File "matplotlib\backends\backend_agg.py", line 496, in print_png
  File "matplotlib\backends\backend_agg.py", line 444, in _print_pil
  File "matplotlib\backends\backend_agg.py", line 387, in draw
  File "matplotlib\artist.py", line 95, in draw_wrapper
  File "matplotlib\artist.py", line 72, in draw_wrapper
  File "matplotlib\figure.py", line 3155, in draw
  File "matplotlib\image.py", line 132, in _draw_list_compositing_images
  File "matplotlib\artist.py", line 72, in draw_wrapper
  File "matplotlib\axes\_base.py", line 3109, in draw
  File "matplotlib\image.py", line 132, in _draw_list_compositing_images
  File "matplotlib\artist.py", line 72, in draw_wrapper
  File "matplotlib\text.py", line 748, in draw
  File "matplotlib\text.py", line 373, in _get_layout
  File "matplotlib\text.py", line 69, in _get_text_metrics_with_cache
  File "matplotlib\text.py", line 77, in _get_text_metrics_with_cache_impl
  File "matplotlib\backends\backend_agg.py", line 219, in get_text_width_height_descent
  File "matplotlib\backends\backend_agg.py", line 256, in _prepare_font
RuntimeError: In set_size: Could not set the fontsize (invalid ppem value; error code 0x97)
```

I understand that your intention might not be to guard against every possible way a user could cause the export to fail, but I did just want to mention it in the event you wanted to clamp the dpi value by some minimum constant. Quick trial-and-error testing reveals that `4` is the minimum acceptable value before exceptions are thrown.
